Adding quickstart guide and organizing LSIF docs (#6642)
sourcegraph/sourcegraph: doc/user/code_intelligence/lsif_quickstart.md

1. Install the [Sourcegraph CLI (`src`)](https://github.com/sourcegraph/src-cli) - used for uploading LSIF data to your Sourcegraph instance.
1. Install the LSIF indexer for your repository's language:
     - Go to https://lsif.dev
     - Find the LSIF indexer for your language
     - Install the indexer as a command-line tool using the installation instructions in the indexer's README

### What is an LSIF indexer?

An LSIF indexer is a command line tool that analyzes your project's source code and generates a file in LSIF format containing all the definitions, references, and hover documentation in your project. That LSIF file is later uploaded to Sourcegraph to provide code intelligence.

## 2. Generate the data

You now need to generate the LSIF data for your repository. Each language's LSIF is unique to that language, so run the command in the _generate LSIF data_ step found in the README of the installed indexer.

## 3. Upload the data

For all languages, the upload step is the same. Make sure the current working directory is somewhere inside your repository, then use the Sourcegraph CLI to run:

```bash
$ src \
  -endpoint=https://sourcegraph.example.com \
  lsif upload \

>   -github-token=abc... (only needed when uploading to Sourcegraph.com) \

Reviewer: I would remove this to keep things simple since it is covered in detail in the *Authentication* section

Author: ```suggestion
  -github-token=<token> \
```

  -file=<LSIF file (e.g. ./cmd/dump.lsif)>
```

If successful, you'll see the following message:

> > Upload successful, queued for processing.

Reviewer: How do I know when the LSIF upload has been processed?

Reviewer 2: Here's what the output looks like:

```
LSIF dump successfully uploaded. It will be converted asynchronously.
To check the status, visit http://localhost:3080/site-admin/lsif-jobs/TFNJRkpvYjoiNTE3OSI=.
```

Author: ```suggestion
> LSIF dump successfully uploaded. It will be converted asynchronously.
> To check the status, visit <link to your Sourcegraph instance LSIF status>
```

> If an error occurred, you'll see it in the response.

Reviewer: What errors could occur?

Reviewer 2: Some errors will appear in the response, others are only shown on the `/lsif-jobs/...` endpoint.

Author: @chrismwendt - I think what we'd like to add here is a list of things that _could_ go wrong. For example, failed authentication (bad token), invalid command, no file found for the given path, etc. What else could go wrong from the server side of things?

Reviewer 3: Unknown repo (404), invalid commit (404), invalid auth if LSIF enforce auth is set to true will (401 if invalid token or 404 if token cannot be checked with github) will be returned from the frontend. 500s for unexpected things (as usual). If the dump is bad (malformed indexer) or we encounter a problem processing (system level bug, flaky connections, etc) then you'll see the job fail in the UI.

Here's the updated src-cli output, btw:

```
$ src-cli -endpoint http://localhost:3080 lsif upload -repo github.com/go-nacelle/process -commit 1e2e6b62d8a4139e0101433bb1bd08db8bb3fa8f -root / -file nacelle-process.lsif
Repository: github.com/go-nacelle/process
Commit: 1e2e6b62d8a4139e0101433bb1bd08db8bb3fa8f
File: nacelle-process.lsif
Root: /

LSIF dump successfully uploaded. It will be converted asynchronously.
To check the status, visit http://localhost:3080/site-admin/lsif-jobs/TFNJRkpvYjoiMTEzIg==.
```

### Authentication

If you're uploading to Sourcegraph.com, you must authenticate your upload by passing a GitHub access token with [`public_repo` scope](https://developer.github.com/apps/building-oauth-apps/understanding-scopes-for-oauth-apps/#available-scopes) as `-github-token=abc...`. You can create one at https://github.com/settings/tokens

## 4. Test out code intelligence

Make sure you have [enabled LSIF code intelligence](lsif.md#enabling-lsif-on-your-sourcegraph-instance) on your Sourcegraph instance.

> Once you have successfully uploaded your LSIF data file to your Sourcegraph instance, go to your repository on Sourcegraph. You should now be able to see precise code intelligence hover tooltips on all files of the language you generated an LSIF index of.

Reviewer 2: ```suggestion
Once you have successfully uploaded your LSIF data file to your Sourcegraph instance, go to your repository on Sourcegraph. You should now be able to see precise code intelligence hover tooltips, definitions, and references on files that were part of the LSIF dump.
```

> Once you have successfully uploaded your LSIF data file to your Sourcegraph instance, go to your repository on Sourcegraph. You should now be able to see precise code intelligence hover tooltips on all files of the language you generated an LSIF index of.

Reviewer: ```suggestion
Once the LSIF data is uploaded, navigate to a code file for the targeted language in the repository, or repository sub-directory LSIF was generated from. LSIF data should now be the source of hover-tooltip content and definitions for that file (presuming that LSIF data exists for that file).

To verify that code intelligence is coming from LSIF:

1. Open your browser's Developer Tools
1. Click on the *Network* tab
1. Reload the page to see all network requests logged
1. Filter network requests by searching for `lsif`.

> NOTE: We are investigating how to make it easier to verify code intelligence is coming from LSIF

```

## 5. Productionize the process

> Now that you're happy with the code intelligence on your repository, you need to make sure that is stays up to date with your repository. This can be done by periodically generating LSIF data, and pushing it to Sourcegraph. You can either [add a step to your CI](./lsif_in_ci.md), or run it as a [GitHub Action](./lsif_on_github.md)

Reviewer: Remove `./`

Author: ```suggestion
Now that you're happy with the code intelligence on your repository, you need to make sure that is stays up to date with your repository. This can be done by periodically generating LSIF data, and pushing it to Sourcegraph. You can either [add a step to your CI](lsif_in_ci.md), or run it as a [GitHub Action](lsif_on_github.md)
```

  lsif upload \
  -github-token=<token> \
  -file=<LSIF file (e.g. ./cmd/dump.lsif)>
```

If successful, you'll see the following message:

> LSIF dump successfully uploaded. It will be converted asynchronously.
> To check the status, visit <link to your Sourcegraph instance LSIF status>

If an error occurred, you'll see it in the response.

### Authentication

If you're uploading to Sourcegraph.com, you must authenticate your upload by passing a GitHub access token with [`public_repo` scope](https://developer.github.com/apps/building-oauth-apps/understanding-scopes-for-oauth-apps/#available-scopes) as `-github-token=abc...`. You can create one at https://github.com/settings/tokens

## 4. Test out code intelligence

Make sure you have [enabled LSIF code intelligence](lsif.md#enabling-lsif-on-your-sourcegraph-instance) on your Sourcegraph instance.

Once the LSIF data is uploaded, navigate to a code file for the targeted language in the repository, or repository sub-directory LSIF was generated from. LSIF data should now be the source of hover-tooltips, definitions, and references for that file (presuming that LSIF data exists for that file).

> To verify that code intelligence is coming from LSIF:

Reviewer 4: Maybe it would be easier if the LSIF handler in the extension logged a statement "Using LSIF code intelligence" on init

## 3. Upload the data

For all languages, the upload step is the same. Make sure the current working directory is somewhere inside your repository, then use the Sourcegraph CLI to run:

```bash
$ src \
  -endpoint=https://sourcegraph.example.com \
  lsif upload \
  -github-token=<token> \
  -file=<LSIF file (e.g. ./cmd/dump.lsif)>
```

If successful, you'll see the following message:

> LSIF dump successfully uploaded. It will be converted asynchronously.
> To check the status, visit <link to your Sourcegraph instance LSIF status>

If an error occurred, you'll see it in the response.

### Authentication

If you're uploading to Sourcegraph.com, you must authenticate your upload by passing a GitHub access token with [`public_repo` scope](https://developer.github.com/apps/building-oauth-apps/understanding-scopes-for-oauth-apps/#available-scopes) as `-github-token=abc...`. You can create one at https://github.com/settings/tokens

## 4. Test out code intelligence

Make sure you have [enabled LSIF code intelligence](lsif.md#enabling-lsif-on-your-sourcegraph-instance) on your Sourcegraph instance.

> Once the LSIF data is uploaded, navigate to a code file for the targeted language in the repository, or repository sub-directory LSIF was generated from. LSIF data should now be the source of hover-tooltip content and definitions for that file (presuming that LSIF data exists for that file).

Author: ```suggestion
Once the LSIF data is uploaded, navigate to a code file for the targeted language in the repository, or repository sub-directory LSIF was generated from. LSIF data should now be the source of hover-tooltips, definitions, and references for that file (presuming that LSIF data exists for that file).
```

To verify that code intelligence is coming from LSIF:

1. Open your browser's Developer Tools
1. Click on the *Network* tab
1. Reload the page to see all network requests logged
1. Filter network requests by searching for `lsif`.

> NOTE: We are investigating how to make it easier to verify code intelligence is coming from LSIF

> 

Author: ```suggestion
```
